### PR TITLE
Add a configurable provider

### LINKED ISSUE #179: Explore possibility of configuration driven Provider for to the letter OAuth2 spec. custom servers
Ideally support some workflow like:

```php
$provider = new League\OAuth2\Client\Provider\Provider([
    'grantType' => 'zzzzzzz', // possibly not needed?
    'tokenUrl' => 'ccccc',
    'refreshUrl' => 'aaaaaa',
    'clientId'  =>  'XXXXXXXX',
    'clientSecret'  =>  'XXXXXXXX',
    'redirectUri'   =>  'https://your-registered-redirect-uri/',
    'scopes' => array('email', '...', '...'),
));
```

### PR DESCRIPTION
**Pull request by shadowhand**
Allows creating providers without having to implement a provider class.

Fixes #179.

### CODE REVIEW

**Comment by shadowhand** [comment]
@ramsey I haven't added tests... is this what you wanted?

File: src/Provider/StandardProvider.php [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+namespace League\OAuth2\Client\Provider;
+
+use InvalidArgumentException;
+
+class StandardProvider extends AbstractProvider
+{
+    /**
+     * @var string
+     */
+    private $urlAuthorize;
```
Comment (shadowhand) [inline src/Provider/StandardProvider.php]: I don't usually like `private`, but I think it makes sense in this case... all the configuration is external and should not be modified after the fact. This keeps it as immutable as an extended class.

File: src/Provider/StandardUser.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        return $this->response[$this->uid];
+    }
+
+    /**
+     * Get the raw user response.
+     *
+     * @return array
+     */
+    public function toArray()
```
Comment (shadowhand) [inline src/Provider/StandardUser.php]: I didn't know what else to do about additional user details...

File: src/Provider/StandardProvider.php (hunk 1 above)
Comment (ramsey) [inline src/Provider/StandardProvider.php]: Makes sense.

File: src/Provider/StandardUser.php (hunk 2 above)
Comment (ramsey) [inline src/Provider/StandardUser.php]: We really don't know what additional user details there will be for a generic provider. User details aren't part of OAuth 2.0, so this is perfect, I think.

**Comment by ramsey** [comment]
This looks like I want was envisioning. Please proceed with tests, and I'll merge it in. Thanks!

**Comment by ramsey** [comment]
I'm curious about the state of this pull request. Are you still working on it?

**Comment by shadowhand** [comment]
@ramsey I put it on hold while #315 gets sorted out. I can probably finish it up tonight.

**Comment by ramsey** [comment]
No worries. Just making sure I didn't miss something. :-)

**Comment by ramsey** [comment]
I deleted the 1.0 branch, promoting dev-master to 1.0.x-dev. Please reopen this pull request against the master branch.

Sorry for the inconvenience.

**Comment by shadowhand** [comment]
@ramsey I'll make a new PR for this after the churn with method names/consistency is worked out.

**Pull request closed without merging**